overlord/devicestate/devicestate.go

```diff
@@ -446,32 +468,38 @@ type pathSideInfo struct {
 }
 
 func (ro *remodelVariant) UpdateWithDeviceContext(st *state.State,
-	pathSI *pathSideInfo, snapName, channel string,
+	pathSI *pathSideInfo, snapName, channel string, revision snap.Revision,
 	userID int, snapStateFlags snapstate.Flags,
 	deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
 	logger.Debugf("snap %s track changed", snapName)
 	if ro.localSnapsRequired {
 		if pathSI == nil {
 			return nil, fmt.Errorf("no snap file provided for %q (track changed)", snapName)
 		}
+		if !revision.Unset() && pathSI.localSi.Revision != revision {
+			return nil, fmt.Errorf("cannot install local snap of different revision: %v != %v", revision, pathSI.localSi.Revision)
+		}
 		return snapstateUpdatePathWithDeviceContext(st, pathSI.localSi, pathSI.path, snapName,
-			&snapstate.RevisionOptions{Channel: channel},
+			&snapstate.RevisionOptions{Channel: channel, Revision: revision},
```

Comment:
until the store relaxes some requirements (which might happen at some point but not immediately) we also need to also pass the validation set(s) this revision was inferred from, same for the other places we are requesting a revision from the store

Reply:
This has been fixed.